File: frontend/src/services/quotes/normalize.js
// Canonical scraper quote:
// { departure_time, arrival_time, duration, price, source, destination }
// Extra keys (airline, collected_at, …) are kept and ignored until the UI needs them.

export function parsePrice(price) {
  if (typeof price === 'number' && Number.isFinite(price)) return price
  const n = Number(String(price ?? '').replace(/[₹Rs.\s,]/gi, ''))
  return Number.isFinite(n) ? n : 0
}

export function parseDurationMinutes(duration) {
  if (typeof duration === 'number') return duration
  const text = String(duration ?? '')
  const h = text.match(/(\d+)\s*h/i)
  const m = text.match(/(\d+)\s*m/i)
  return (h ? Number(h[1]) : 0) * 60 + (m ? Number(m[1]) : 0)
}

export function formatInr(n) {
  return `₹${Number(n || 0).toLocaleString('en-IN')}`
}

export function normalizeQuote(raw, index = 0) {
  return {
    id: raw.id ?? `Q-${String(index + 1).padStart(4, '0')}`,
    departure_time: raw.departure_time ?? '',
    arrival_time: raw.arrival_time ?? '',
    duration: raw.duration ?? '',
    duration_minutes: raw.duration_minutes ?? parseDurationMinutes(raw.duration),
    price: raw.price ?? '',
    price_inr: raw.price_inr ?? parsePrice(raw.price),
    source: raw.source ?? '',
    destination: raw.destination ?? '',
    collected_at: raw.collected_at ?? null,
    airline: raw.airline ?? null,
  }
}

export function routeKey(source, destination) {
  return `${source} → ${destination}`
}

File: backend/app/services/normalize.py
from __future__ import annotations

import hashlib
import re
from typing import Any

from pydantic import ValidationError

from app.models.quote import Quote

_DURATION_H = re.compile(r"(\d+)\s*h", re.IGNORECASE)
_DURATION_M = re.compile(r"(\d+)\s*m", re.IGNORECASE)
_PRICE_CHARS = re.compile(r"[^\d.]")


def parse_price_inr(price: Any) -> int | None:
    if isinstance(price, bool):
        return None
    if isinstance(price, (int, float)) and price >= 0:
        return int(price)
    text = _PRICE_CHARS.sub("", str(price or ""))
    if not text:
        return None
    try:
        return int(float(text))
    except ValueError:
        return None


def parse_duration_minutes(duration: Any) -> int:
    if isinstance(duration, bool):
        return 0
    if isinstance(duration, (int, float)) and duration >= 0:
        return int(duration)
    text = str(duration or "")
    hours = _DURATION_H.search(text)
    minutes = _DURATION_M.search(text)
    return (int(hours.group(1)) if hours else 0) * 60 + (int(minutes.group(1)) if minutes else 0)


def _stable_id(parts: list[str]) -> str:
    digest = hashlib.sha256("|".join(parts).encode("utf-8")).hexdigest()[:12]
    return f"Q-{digest}"


def _as_text(value: Any) -> str:
    if value is None:
        return ""
    return str(value).strip()


def normalize_raw_quote(raw: dict[str, Any]) -> Quote | None:
    source = _as_text(raw.get("source"))
    destination = _as_text(raw.get("destination"))
    departure_time = _as_text(raw.get("departure_time"))
    price = _as_text(raw.get("price"))
    price_inr = parse_price_inr(raw.get("price"))

    if not source or not destination or not departure_time or not price or price_inr is None:
        return None

    duration = _as_text(raw.get("duration"))
    travel_date = _as_text(raw.get("date"))
    scraped_on = _as_text(raw.get("today") or raw.get("collected_at"))
    airline = _as_text(raw.get("airline") or raw.get("airline_name")) or None

    quote_id = _as_text(raw.get("id")) or _stable_id(
        [
            source,
            destination,
            travel_date,
            departure_time,
            _as_text(raw.get("arrival_time")),
            price,
            airline or "",
            _as_text(raw.get("plane_number")),
        ]
    )

    payload = {
        "id": quote_id,
        "departure_time": departure_time,
        "arrival_time": _as_text(raw.get("arrival_time")),
        "duration": duration,
        "duration_minutes": parse_duration_minutes(duration),
        "price": price,
        "price_inr": price_inr,
        "source": source,
        "destination": destination,
        "collected_at": scraped_on or None,
        "travel_date": travel_date or None,
        "airline": airline,
    }

    try:
        return Quote.model_validate(payload)
    except ValidationError:
        return None


def normalize_quotes(raw_rows: list[dict[str, Any]]) -> list[Quote]:
    quotes: list[Quote] = []
    seen: set[str] = set()
    for row in raw_rows:
        quote = normalize_raw_quote(row)
        if quote is None or quote.id in seen:
            continue
        seen.add(quote.id)
        quotes.append(quote)
    return quotes
